drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1568 @@
+		glBufferData(GL_ARRAY_BUFFER, sizeof(BatchVertexColored) * bdata.vertices_colored.size(), bdata.vertices_colored.get_data(), GL_DYNAMIC_DRAW);
 	}
 
-	state.uniforms.projection_matrix = canvas_transform;
-
-	state.uniforms.final_modulate = Color(1, 1, 1, 1);
-
-	state.uniforms.modelview_matrix = Transform2D();
-	state.uniforms.extra_matrix = Transform2D();
-
-	_set_uniforms();
-	_bind_quad_buffer();
+	// might not be necessary
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
 }
 
-void RasterizerCanvasGLES2::canvas_end() {
+RasterizerCanvasGLES2::Batch *RasterizerCanvasGLES2::_batch_request_new(bool p_blank) {
+	Batch *batch = bdata.batches.request();
+	if (!batch) {
+		// grow the batches
+		bdata.batches.grow();
 
-	glBindBuffer(GL_ARRAY_BUFFER, 0);
+		// and the temporary batches (used for color verts)
+		bdata.batches_temp.reset();
+		bdata.batches_temp.grow();
 
-	for (int i = 0; i < VS::ARRAY_MAX; i++) {
-		glDisableVertexAttribArray(i);
+		// this should always succeed after growing
+		batch = bdata.batches.request();
+#ifdef DEBUG_ENABLED
+		CRASH_COND(!batch);
+#endif
 	}
 
-	if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_DIRECT_TO_SCREEN]) {
-		//reset viewport to full window size
-		int viewport_width = OS::get_singleton()->get_window_size().width;
-		int viewport_height = OS::get_singleton()->get_window_size().height;
-		glViewport(0, 0, viewport_width, viewport_height);
-		glScissor(0, 0, viewport_width, viewport_height);
-	}
+	if (p_blank)
+		memset(batch, 0, sizeof(Batch));
 
-	state.using_texture_rect = false;
-	state.using_skeleton = false;
-	state.using_ninepatch = false;
-	state.using_transparent_rt = false;
+	return batch;
 }
 
-RasterizerStorageGLES2::Texture *RasterizerCanvasGLES2::_bind_canvas_texture(const RID &p_texture, const RID &p_normal_map) {
+// This function may be called MULTIPLE TIMES for each item, so needs to record how far it has got
+bool RasterizerCanvasGLES2::prefill_joined_item(FillState &r_fill_state, int &r_command_start, Item *p_item, Item *p_current_clip, bool &r_reclip, RasterizerStorageGLES2::Material *p_material) {
+	// we will prefill batches and vertices ready for sending in one go to the vertex buffer
+	int command_count = p_item->commands.size();
+	Item::Command *const *commands = p_item->commands.ptr();
 
-	RasterizerStorageGLES2::Texture *tex_return = NULL;
+	Transform2D transform;
+	TransformMode transform_mode = _find_transform_mode(r_fill_state.use_hardware_transform, p_item->final_transform, transform);
 
-	if (p_texture.is_valid()) {
+	Vector2 texpixel_size = r_fill_state.texpixel_size;
 
-		RasterizerStorageGLES2::Texture *texture = storage->texture_owner.getornull(p_texture);
+	// checking the color for not being white makes it 92/90 times faster in the case where it is white
+	bool multiply_final_modulate = false;
+	if (!r_fill_state.use_hardware_transform && (r_fill_state.final_modulate != Color(1, 1, 1, 1))) {
```

Comment:
"!= Color()" breaks down to a series of floating point "!=" comparisons which can have issues. It is likely best to use ``is_equal_approx()``
```suggestion
	if (!r_fill_state.use_hardware_transform && (!r_fill_state.final_modulate.is_equal_approx(Color(1, 1, 1, 1)))) {
```